apps/recnet-api/src/modules/invite-code/invite-code.service.ts

```diff
@@ -6,37 +6,68 @@ import {
   InviteCode as DbInviteCode,
   InviteCodeFilterBy,
 } from "@recnet-api/database/repository/invite-code.repository.type";
+import UserRepository from "@recnet-api/database/repository/user.repository";
 import { getOffset } from "@recnet-api/utils";
 
 import { InviteCode } from "./entities/invite-code.entity";
 import {
   CreateInviteCodeResponse,
-  GetInviteCodeResponse,
+  GetAllInviteCodeResponse,
 } from "./invite-code.response";
 
 @Injectable()
 export class InviteCodeService {
   constructor(
     @Inject(InviteCodeRepository)
-    private readonly inviteCodeRepository: InviteCodeRepository
+    private readonly inviteCodeRepository: InviteCodeRepository,
+    @Inject(UserRepository)
+    private readonly userRepository: UserRepository
   ) {}
 
   public async createInviteCode(
     numCodes: number,
-    ownerId: string
+    ownerId: string | null,
+    upperBound: number | null
   ): Promise<CreateInviteCodeResponse> {
-    const codes = Array.from({ length: numCodes }, () => this.genRandomCode());
-    await this.inviteCodeRepository.createInviteCode(codes, ownerId);
+    const targetUserIds: string[] = [];
+    if (ownerId) {
+      targetUserIds.push(ownerId);
+    } else {
+      const users = await this.userRepository.findAllUsers();
+      targetUserIds.push(...users.map((user) => user.id));
+    }
+
+    const codeOwnerPairsArr = await Promise.all(
+      targetUserIds.map(async (userId) => {
+        // calculate how many codes this user already has
+        const numUserCodes = await this.inviteCodeRepository.countInviteCodes({
```

Comment:
It will be better if we use groupBy to query the existing count for all users as it will only DB I/O once.
Also, we should be careful when using `...` if the array is large because it will copy the array every time and reallocate the memory. Performance might be a concern when the number of user grows.

I rewrote a version that I thought was more efficient for your reference. 
```ts
// invite-codes.repository.ts
public async countInviteCodesByOwnerIds(
    ownerIds: string[]
): Promise<Array<{ ownerId: string; count: number }>> {
    const result = await this.prisma.inviteCode.groupBy({
      by: ["ownerId"],
      where: { ownerId: { in: ownerIds } },
      _count: { code: true },
    });
    return result.map((r) => ({ ownerId: r.ownerId, count: r._count.code }));
}

// invite-codes.service.ts
public async createInviteCode(
    numCodes: number,
    ownerId: string | null,
    upperBound: number | null
): Promise<CreateInviteCodeResponse> {
    const targetUserIds: string[] = [];
    if (ownerId) {
      targetUserIds.push(ownerId);
    } else {
      const users = await this.userRepository.findAllUsers();
      targetUserIds.push(...users.map((user) => user.id));
    }

    const codesToBeCreated: {
      code: string;
      ownerId: string;
    }[] = [];

    const userExistingCodesCount =
      await this.inviteCodeRepository.countInviteCodesByOwnerIds(targetUserIds);

    userExistingCodesCount.forEach(({ ownerId, count: existingCodeCount }) => {
      const numCodesToGenerate =
        upperBound !== null
          ? Math.min(upperBound - existingCodeCount, numCodes)
          : numCodes;

      const newCodes = Array.from({ length: numCodesToGenerate }, () => ({
        ownerId,
        code: this.genRandomCode(),
      }));
      codesToBeCreated.push(...newCodes);
    });

    await this.inviteCodeRepository.createInviteCode(codesToBeCreated);
    return {
      codes: codesToBeCreated.map((pair) => pair.code),
    };
}
```

Reply:
Good catch about the performance aspect! Fixed.